### PR TITLE
New NETCONF get/get-config module and adding NETCONF support for Nokia SROS devices

### PR DESCRIPTION
**Pull request by wisotzky**
##### SUMMARY
Support SROS 16.0 devices when used with NETCONF.
New NETCONF module to do get/get-config.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
- NETCONF plugin
- NETCONF module netconf_get

##### ANSIBLE VERSION

```
2.5.0
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/plugins/netconf/sros.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+        xmlns = "urn:nokia.com:sros:ns:yang:sr:state"
+        f = '<state xmlns="%s"><system><platform/><bootup/><version/><lldp/></system></state>' % xmlns
+        reply = to_ele(self.m.get(filter=('subtree', f)).data_xml)
+
+        device_info['network_os_hostname'] = reply.findtext('.//{%s}state/{*}system/{*}lldp/{*}system-name' % xmlns)
+        device_info['network_os_version'] = reply.findtext('.//{%s}state/{*}system/{*}version/{*}version-number' % xmlns)
+        device_info['network_os_model'] = reply.findtext('.//{%s}state/{*}system/{*}platform' % xmlns)
+
+        # device_info['network_os_platform'] = 'unknown'
```
Comment (ganeshrn) [inline lib/ansible/plugins/netconf/sros.py]: Is this commented line required? If not it can be removed.

Comment (wisotzky) [reply lib/ansible/plugins/netconf/sros.py]: I've checked, what device-info other vendor implementations do support. I've added those attributes as comments just as place-holder, until I do understand what a good mapping for those would be. Unfortunately I currently have no access to IOSXR or JUNOS boxes, to give it a try.
Is there any good guide, which describes what device_info entries shall be supported and how they typically should be mapped. Example could even be, that some static mapping is applied, so source of info is not required to be NETCONF.

**Comment by wisotzky** [comment]
I've intended to open a new pull request containing just the new netconf module netconf_get. Unfortunately, it has combined it with my previous changes for better SROS support in the same shipment. Any way to get this two deliverables separator - or just okay for now?

**Comment by wisotzky** [comment]
Closed/Reopened to run CI again.

**Comment by wisotzky** [comment]
retested connection=local and connection=netconf

**Review by ganeshrn — CHANGES_REQUESTED** [review]
`netconf_get` module is added as part of PR https://github.com/ansible/ansible/pull/39869 as per the proposal https://github.com/ansible/proposals/issues/104 which was discussed and agreed in `ansible-network` irc meeting. The above PR has most of the functionality added here with some deviations. 
1) As `connection=local` will be deprecated for network modules it was decided that going ahead all-new network modules over netconf will support only `connection=netconf`.
2) The post filter part is not added in module as it can be achieved using other modules in Ansible eg: `xml` module.

**Comment by wisotzky** [comment]
@ganeshrn, I am getting confused about the content of PR #39869. I've invested quite some time in the creation of that Ansible module including testing. So why do you rewrite almost my complete contribution and call yourself the author? I don't think this is the intention of open-source. Are you paid by the number of Ansible modules you authored? If such case, I will stop immediate to contribute python code to Ansible - and I will start to push my requests just as feature-requests and bug reports. 

About the content:
* You need to change "confgiuration" to "configuration" and some more typos.
* The option "source=auto" does not make any sense for me. Most of the users will use "source=running" or "source=None". There is no such use-case to access "candidate" and fallback to "running" if not supported. The only reason to access the "candidate" is if you interested to review uncommitted changes - but in such case you would explicit ask for candidate.
* The post-filtering you've removed was mainly to use XPATH filters on devices, which don't support the :xpath capability. So having the same, shorter and more intuitive syntax. While your comment on using the XML module might be right - it means the playbook designer need to take care about what the device actually support to write the playbook differently - or he just could always fallback to subtree filters. The other option actually would be, to translate simple xpath filter statement into a subtree statement to do pre-filtering...
* You've introduced the option "display=json". Following your argument on post-filtering, I would actually say that this functionality could easily sit in another Ansible module. Especially as you always return the raw XML result. My intend was to let the user decide, how the result should be rendered - so ideally in XML, JSON or YAML format.
* The example section should contain one example, how to do a <get> with a filter. Also it would be reasonable to add one XPATH filtering example. Last but not least, your 3rd example would not work on most systems, as it missed the XML namespace for the <configuration> XML element.

**Comment by wisotzky** [comment]
@ganeshrn, what happens to the other changes for better SROS NETCONF support in that PR?

**Comment by ganeshrn** [comment]
@wisotzky The first step while adding a feature to the open-source project is to figure out if the feature already exists and if not is someone else working on that feature?

If you had taken the effort to find that out you would have come across links [1] and [2] and that would have helped avoid the effort duplication. This is one of the main reasons for Github issue tracking, IRC based community meeting, public wiki roadmaps.

[1] Github issue: https://github.com/ansible/ansible/issues/38819
[2] Proposal: https://github.com/ansible/proposals/issues/104

The proposal was discussed in weekly IRC meeting [3] and [4]. We had asked community members to review the proposal and it was agreed upon unanimously by the attendees with 2.6 target release. This was done way before you added `netconf_get` module in this PR.

[3] https://github.com/ansible/community/issues/247#issuecomment-380866634
[4] IRC meeting details: https://github.com/ansible/community/tree/master/group-network

Also, FYI I don't get paid based on the number of Ansible modules I author. On the contrary, I am happy to have more community contributions. Based on your comments I assume that you don't attend `ansible-network` irc community meetings and are not in sync with the latest development in the project. Had you mentioned either in issue comment or proposal that you are working on adding a similar feature I could have happily stopped working on this and moved on to work on other important items in the roadmap that needs to be done for 2.6 release.

With regards to the PR that I raised the module capability is more or less similar to that in this PR which is also similar to what is in the proposal but the code flow is entirely different as I suggested in my earlier review comments. It is **not** a rewrite of the module in this PR.

**Comment by ganeshrn** [comment]
@wisotzky The `netconf_get` module added in this PR in it's current form is not acceptable.

If you are ok I can add you as `netconf_get` module co-author and maintainer in PR #39869 and we can collaborate on `netconf_get` effort in that PR. 

For this PR I suggest you to revert the netconf_get changes and have this PR dedicated for `sros` related code.

**Comment by wisotzky** [comment]
@ganeshrn Actually I've not yet part of that IRC channel - and don't even know, if I've got cycles enough to be deeply involved in the planning process. About the netconf_get module, quite a coincidence that we've started to work on this almost at the very same time - and that some of the ideas like having a common filter attribute for subtree/xpath and JSON output are contained in both contributions.

Could you please check on the other comments I've made above? I have not clue how to create a fork of your PR, to do the corrections by myself.

**Comment by ganeshrn** [comment]
@wisotzky SInce the module is based on a standard ietf netconf rfc imo it is quite possible to have similar functionality while implementing this feature.
 
We have `junos_command` module providing output in json format as xml is not natively consumed within Ansible hence it make sense to add it in netconf_get as well.

Below is one possible way to add changes in a netconf_get branch for PR #39869
```
$ git clone https://github.com/wisotzky/ansible
$ cd ansible
$ git remote add upstream https://github.com/ganeshrn/ansible
$ git remote -v
$ git fetch upstream
$ git checkout -b netconf_get
$ git reset --hard upstream/netconf_get
$ git push origin netconf_get
```
Make the changes locally and push it in your repo branch
```
$ git add <changes>
$ git commit
$ git push origin netconf_get
```
After that raise a PR with `netconf_get` branch selected in Github UI.

I am available in #ansible-network irc channel by nick @ganeshrn.
If you need any help or want to discuss more on this, please reach out to me on irc.

**Comment by ganeshrn** [comment]
>You need to change "confgiuration" to "configuration" and some more typos.

Will do.

>The option "source=auto" does not make any sense for me. Most of the users will use "source=running" or "source=None". There is no such use-case to access "candidate" and fallback to "running" if not supported. The only reason to access the "candidate" is if you interested to review uncommitted changes - but in such case you would explicit ask for candidate.

Agree on this, having `auto` option might add to the confusion, will modify it to support `running`, `candidate` and `startup` with `running` as default.

>The post-filtering you've removed was mainly to use XPATH filters on devices, which don't support the :xpath capability. So having the same, shorter and more intuitive syntax. While your comment on using the XML module might be right - it means the playbook designer need to take care about what the device actually support to write the playbook differently - or he just could always fallback to subtree filters. The other option actually would be, to translate simple xpath filter statement into a subtree statement to do pre-filtering...

`xml` module with `xpath` capability is readily available in Ansible, adding xpath to subtree magic in module will make the module more complicated and confusing for users. Based on the experience with other network modules we figured out that implementation not only varies across vendors but also across different product lines of the same vendor and sometimes across different versions of the same OS. Thus adding logic within module that is not supported natively by network host results in complex code path over time making module more prone to errors. That's we decided to report an appropriate error in case the feature is not supported natively so that user can take appropriate corrective action.

>You've introduced the option "display=json". Following your argument on post-filtering, I would actually say that this functionality could easily sit in another Ansible module. Especially as you always return the raw XML result. My intend was to let the user decide, how the result should be rendered - so ideally in XML, JSON or YAML format.

Converting `xml` to `json` is not supported directly within Ansible, to achieve this user might require to write custom filter and is not as straightforward as using `xml` module (as mentioned in above command). Also, the `json` output is made optional as it requires installing third part library (jxmlease).

>The example section should contain one example, how to do a with a filter. Also it would be reasonable to add one XPATH filtering example. Last but not least, your 3rd example would not work on most systems, as it missed the XML namespace for the XML element.

If you have an example with xpath filter and namespace readily available can you please add it to the 
PR. The 3rd example is tested and works with junos platform, I will add more examples tested with `iosxr` and `nxos` platforms.

**Pull request closed without merging**